### ISSUE TITLE
Single pod termination in Statefulset causes all lower ordinal pods to terminate which can cause yellow state if highest ordinal pod is terminated

### ISSUE THREAD

**Issue by ccben87** [Opster/opensearch-k8s-operator#464]
Single pod termination in Statefulset causes all lower ordinal pods to terminate which can cause cluster yellow status if the original terminated pod is the highest ordinal in the Statefulset which results in termination of all pods in the node group. I have found that this happens on version 2.4.1 and onwards but not on 2.4.0 and earlier. When testing by terminating a pod in a cluster of 3 nodes, all pods with a lower ordinal number terminate (all pods in the node group terminate) which causes a yellow cluster state. I have also found that disabling the dashboard results in an odd behaviour in versions 2.4.0 and below when terminating the highest ordinal pod in a node group's Statefulset. When this is done in a Statefulset with 3 replicas, the middle ordinal numbered pod gets terminated as well as the highest ordinal pod number but the lowest ordinal number pod remains up and working properly. Not sure why changing dashboard from true to false would cause the middle ordinal pod in a Statefulset of 3 replicas to incorrectly terminate when the highest ordinal pod gets terminated with the lowest ordinal working normally.

I can replicate this using the default config provided in your guide with the version number increased to 2.4.1:

```yaml
apiVersion: opensearch.opster.io/v1
kind: OpenSearchCluster
metadata:
  name: my-first-cluster
  namespace: default
spec:
  general:
    serviceName: my-first-cluster
    version: 2.4.1
  dashboards:
    enable: true
    version: 2.4.1
    replicas: 1
    resources:
      requests:
         memory: "512Mi"
         cpu: "200m"
      limits:
         memory: "512Mi"
         cpu: "200m"
  nodePools:
    - component: nodes
      replicas: 3
      diskSize: "5Gi"
      nodeSelector:
      resources:
         requests:
            memory: "2Gi"
            cpu: "500m"
         limits:
            memory: "2Gi"
            cpu: "500m"
      roles:
        - "cluster_manager"
        - "data"
```

With the above configuration (version 2.4.1), when running "kubectl delete pod/my-first-cluster-nodes-2" the following happens:

![image](https://user-images.githubusercontent.com/128038118/225584805-8a90a239-ddc9-4628-b142-ab65f9e336df.png)

With the same configuration but with version 2.4.0 this happens (which is expected):

![image](https://user-images.githubusercontent.com/128038118/225587242-3c5e5de5-3cd0-4999-883e-a95a8ade7470.png)

This is what happens when running "kubectl delete pod/my-first-cluster-nodes-1", you can see that only the lower ordinal numbered pods get terminated.

![image](https://user-images.githubusercontent.com/128038118/225589258-82ff3a24-79ad-4551-bfb0-4d6d08ab0405.png)

I'm not sure if this is a bug in Opensearch or a bug in Opensearch Operator or how to troubleshoot this further.

**Comment by idanl21** [comment]
Hey @ccben87 
Looks really strange, I wonder how's the diffrence between the versions (2.4 to 2.4.1) is related to the issue. It can also be OpenSearch bug, but it also do not sounds reasonable. Did you had a chance to check also other versions?
@swoehrl-mw do you have any thoughts ?

**Comment by mchiappini** [comment]
I can confirm this is happening to us too. Although we see this behaviour on `2.4.0`:
* Operator version `2.2.0`
* Opensearch version: `2.4.0`

**Comment by swoehrl-mw** [comment]
> do you have any thoughts ?

From the experiences described here I'd say this is not a problem in opensearch but rather strange behaviour of kubernetes itself. I'm pretty sure its not the operator deliberately restarting the pods. Maybe the operator sets some statefulset option that makes kubernetes behave that way. 
Unfortunately I don't have the time currently to debug this in depth.

**Comment by ccben87** [comment]
I've been looking at this from a Kubernetes angle and I've looked at the debug information for kubelet and get a "SyncLoop DELETE" source="api" for the pod in question. I then went and looked at the control plane logs for AWS EKS in Cloudwatch and managed to find this:

`{
    "kind": "Event",
    "apiVersion": "audit.k8s.io/v1",
    "level": "RequestResponse",
    "auditID": "9b3289b2-cd4e-493a-9a6a-bf2649875224",
    "stage": "ResponseComplete",
    "requestURI": "/api/v1/namespaces/default/pods/my-first-cluster-nodes-0",
    "verb": "delete",
    "user": {
        "username": "system:serviceaccount:default:opensearch-operator-controller-manager",
        "uid": "7bdd022a-0e07-4224-812e-367f1b4e8a1f",
        "groups": [
            "system:serviceaccounts",
            "system:serviceaccounts:default",
            "system:authenticated"
        ],
        "extra": {
            "authentication.kubernetes.io/pod-name": [
                "opensearch-operator-controller-manager-cbb84746d-5kq89"
            ],
            "authentication.kubernetes.io/pod-uid": [
                "32d5b17c-f738-428d-9446-cb58f2adf4f9"
            ]
        }
    },
    "sourceIPs": [
        "192.168.146.79"
    ],
    "userAgent": "manager/v0.0.0 (linux/arm64) kubernetes/$Format",
    "objectRef": {
        "resource": "pods",
        "namespace": "default",
        "name": "my-first-cluster-nodes-0",
        "apiVersion": "v1"
    },
    "responseStatus": {
        "metadata": {},
        "code": 200
    },
    "requestObject": {
        "kind": "DeleteOptions",
        "apiVersion": "meta.k8s.io/__internal"
    },
    "responseObject": {
        "kind": "Pod",
        "apiVersion": "v1",
        "metadata": {
            "name": "my-first-cluster-nodes-0",
            "generateName": "my-first-cluster-nodes-",
            "namespace": "default",
            "uid": "5ac52fc2-d752-4da7-813d-0ead61cf5011",
            "resourceVersion": "69011",
            "creationTimestamp": "2023-04-19T08:53:38Z",
            "deletionTimestamp": "2023-04-19T09:11:15Z",
            "deletionGracePeriodSeconds": 30,
            "labels": {
                "controller-revision-hash": "my-first-cluster-nodes-6697b4cf78",
                "opensearch.role": "cluster_manager",
                "opster.io/opensearch-cluster": "my-first-cluster",
                "opster.io/opensearch-nodepool": "nodes",
                "statefulset.kubernetes.io/pod-name": "my-first-cluster-nodes-0"
            },
            "annotations": {
                "opster.io/config": ""
            },
            "ownerReferences": [
                {
                    "apiVersion": "apps/v1",
                    "kind": "StatefulSet",
                    "name": "my-first-cluster-nodes",
                    "uid": "1bb11bfa-e1a3-47e5-b3d7-093b5033ccf4",
                    "controller": true,
                    "blockOwnerDeletion": true
                }
            ],
            "managedFields": [
                {
                    "manager": "kube-controller-manager",
                    "operation": "Update",
                    "apiVersion": "v1",
                    "time": "2023-04-19T08:53:38Z",
                    "fieldsType": "FieldsV1",
                    "fieldsV1": {
                        "f:metadata": {
                            "f:annotations": {
                                ".": {},
                                "f:opster.io/config": {}
                            },
                            "f:generateName": {},
                            "f:labels": {
                                ".": {},
                                "f:controller-revision-hash": {},
                                "f:opensearch.role": {},
                                "f:opster.io/opensearch-cluster": {},
                                "f:opster.io/opensearch-nodepool": {},
                                "f:statefulset.kubernetes.io/pod-name": {}
                            },
                            "f:ownerReferences": {
                                ".": {},
                                "k:{\"uid\":\"1bb11bfa-e1a3-47e5-b3d7-093b5033ccf4\"}": {}
                            }
                        },
                        "f:spec": {
                            "f:affinity": {
                                ".": {},
                                "f:podAntiAffinity": {
                                    ".": {},
                                    "f:preferredDuringSchedulingIgnoredDuringExecution": {}
                                }
                            },
                            "f:containers": {
                                "k:{\"name\":\"opensearch\"}": {
                                    ".": {},
                                    "f:command": {},
                                    "f:env": {
                                        ".": {},
                                        "k:{\"name\":\"OPENSEARCH_JAVA_OPTS\"}": {
                                            ".": {},
                                            "f:name": {},
                                            "f:value": {}
                                        },
                                        "k:{\"name\":\"cluster.initial_master_nodes\"}": {
                                            ".": {},
                                            "f:name": {},
                                            "f:value": {}
                                        },
                                        "k:{\"name\":\"cluster.name\"}": {
                                            ".": {},
                                            "f:name": {},
                                            "f:value": {}
                                        },
                                        "k:{\"name\":\"discovery.seed_hosts\"}": {
                                            ".": {},
                                            "f:name": {},
                                            "f:value": {}
                                        },
                                        "k:{\"name\":\"http.port\"}": {
                                            ".": {},
                                            "f:name": {},
                                            "f:value": {}
                                        },
                                        "k:{\"name\":\"network.bind_host\"}": {
                                            ".": {},
                                            "f:name": {},
                                            "f:value": {}
                                        },
                                        "k:{\"name\":\"network.publish_host\"}": {
                                            ".": {},
                                            "f:name": {},
                                            "f:valueFrom": {
                                                ".": {},
                                                "f:fieldRef": {}
                                            }
                                        },
                                        "k:{\"name\":\"node.roles\"}": {
                                            ".": {},
                                            "f:name": {},
                                            "f:value": {}
                                        }
                                    },
                                    "f:image": {},
                                    "f:imagePullPolicy": {},
                                    "f:livenessProbe": {
                                        ".": {},
                                        "f:failureThreshold": {},
                                        "f:initialDelaySeconds": {},
                                        "f:periodSeconds": {},
                                        "f:successThreshold": {},
                                        "f:tcpSocket": {
                                            ".": {},
                                            "f:port": {}
                                        },
                                        "f:timeoutSeconds": {}
                                    },
                                    "f:name": {},
                                    "f:ports": {
                                        ".": {},
                                        "k:{\"containerPort\":9200,\"protocol\":\"TCP\"}": {
                                            ".": {},
                                            "f:containerPort": {},
                                            "f:name": {},
                                            "f:protocol": {}
                                        },
                                        "k:{\"containerPort\":9300,\"protocol\":\"TCP\"}": {
                                            ".": {},
                                            "f:containerPort": {},
                                            "f:name": {},
                                            "f:protocol": {}
                                        }
                                    },
                                    "f:readinessProbe": {
                                        ".": {},
                                        "f:exec": {
                                            ".": {},
                                            "f:command": {}
                                        },
                                        "f:failureThreshold": {},
                                        "f:initialDelaySeconds": {},
                                        "f:periodSeconds": {},
                                        "f:successThreshold": {},
                                        "f:timeoutSeconds": {}
                                    },
                                    "f:resources": {
                                        ".": {},
                                        "f:limits": {
                                            ".": {},
                                            "f:cpu": {},
                                            "f:memory": {}
                                        },
                                        "f:requests": {
                                            ".": {},
                                            "f:cpu": {},
                                            "f:memory": {}
                                        }
                                    },
                                    "f:startupProbe": {
                                        ".": {},
                                        "f:failureThreshold": {},
                                        "f:initialDelaySeconds": {},
                                        "f:periodSeconds": {},
                                        "f:successThreshold": {},
                                        "f:tcpSocket": {
                                            ".": {},
                                            "f:port": {}
                                        },
                                        "f:timeoutSeconds": {}
                                    },
                                    "f:terminationMessagePath": {},
                                    "f:terminationMessagePolicy": {},
                                    "f:volumeMounts": {
                                        ".": {},
                                        "k:{\"mountPath\":\"/mnt/admin-credentials\"}": {
                                            ".": {},
                                            "f:mountPath": {},
                                            "f:name": {}
                                        },
                                        "k:{\"mountPath\":\"/usr/share/opensearch/data\"}": {
                                            ".": {},
                                            "f:mountPath": {},
                                            "f:name": {}
                                        }
                                    }
                                }
                            },
                            "f:dnsPolicy": {},
                            "f:enableServiceLinks": {},
                            "f:hostname": {},
                            "f:initContainers": {
                                ".": {},
                                "k:{\"name\":\"init\"}": {
                                    ".": {},
                                    "f:args": {},
                                    "f:command": {},
                                    "f:image": {},
                                    "f:imagePullPolicy": {},
                                    "f:name": {},
                                    "f:resources": {},
                                    "f:securityContext": {
                                        ".": {},
                                        "f:runAsUser": {}
                                    },
                                    "f:terminationMessagePath": {},
                                    "f:terminationMessagePolicy": {},
                                    "f:volumeMounts": {
                                        ".": {},
                                        "k:{\"mountPath\":\"/usr/share/opensearch/data\"}": {
                                            ".": {},
                                            "f:mountPath": {},
                                            "f:name": {}
                                        }
                                    }
                                }
                            },
                            "f:restartPolicy": {},
                            "f:schedulerName": {},
                            "f:securityContext": {},
                            "f:subdomain": {},
                            "f:terminationGracePeriodSeconds": {},
                            "f:topologySpreadConstraints": {
                                ".": {},
                                "k:{\"topologyKey\":\"kubernetes.io/hostname\",\"whenUnsatisfiable\":\"DoNotSchedule\"}": {
                                    ".": {},
                                    "f:labelSelector": {},
                                    "f:maxSkew": {},
                                    "f:topologyKey": {},
                                    "f:whenUnsatisfiable": {}
                                },
                                "k:{\"topologyKey\":\"topology.kubernetes.io/zone\",\"whenUnsatisfiable\":\"DoNotSchedule\"}": {
                                    ".": {},
                                    "f:labelSelector": {},
                                    "f:maxSkew": {},
                                    "f:topologyKey": {},
                                    "f:whenUnsatisfiable": {}
                                }
                            },
                            "f:volumes": {
                                ".": {},
                                "k:{\"name\":\"admin-credentials\"}": {
                                    ".": {},
                                    "f:name": {},
                                    "f:secret": {
                                        ".": {},
                                        "f:defaultMode": {},
                                        "f:secretName": {}
                                    }
                                },
                                "k:{\"name\":\"data\"}": {
                                    ".": {},
                                    "f:name": {},
                                    "f:persistentVolumeClaim": {
                                        ".": {},
                                        "f:claimName": {}
                                    }
                                }
                            }
                        }
                    }
                },
                {
                    "manager": "kubelet",
                    "operation": "Update",
                    "apiVersion": "v1",
                    "time": "2023-04-19T08:55:24Z",
                    "fieldsType": "FieldsV1",
                    "fieldsV1": {
                        "f:status": {
                            "f:conditions": {
                                "k:{\"type\":\"ContainersReady\"}": {
                                    ".": {},
                                    "f:lastProbeTime": {},
                                    "f:lastTransitionTime": {},
                                    "f:status": {},
                                    "f:type": {}
                                },
                                "k:{\"type\":\"Initialized\"}": {
                                    ".": {},
                                    "f:lastProbeTime": {},
                                    "f:lastTransitionTime": {},
                                    "f:status": {},
                                    "f:type": {}
                                },
                                "k:{\"type\":\"Ready\"}": {
                                    ".": {},
                                    "f:lastProbeTime": {},
                                    "f:lastTransitionTime": {},
                                    "f:status": {},
                                    "f:type": {}
                                }
                            },
                            "f:containerStatuses": {},
                            "f:hostIP": {},
                            "f:initContainerStatuses": {},
                            "f:phase": {},
                            "f:podIP": {},
                            "f:podIPs": {
                                ".": {},
                                "k:{\"ip\":\"192.168.163.58\"}": {
                                    ".": {},
                                    "f:ip": {}
                                }
                            },
                            "f:startTime": {}
                        }
                    },
                    "subresource": "status"
                }
            ]
        },
        "spec": {
            "volumes": [
                {
                    "name": "data",
                    "persistentVolumeClaim": {
                        "claimName": "data-my-first-cluster-nodes-0"
                    }
                },
                {
                    "name": "admin-credentials",
                    "secret": {
                        "secretName": "my-first-cluster-admin-password",
                        "defaultMode": 420
                    }
                },
                {
                    "name": "kube-api-access-h2thl",
                    "projected": {
                        "sources": [
                            {
                                "serviceAccountToken": {
                                    "expirationSeconds": 3607,
                                    "path": "token"
                                }
                            },
                            {
                                "configMap": {
                                    "name": "kube-root-ca.crt",
                                    "items": [
                                        {
                                            "key": "ca.crt",
                                            "path": "ca.crt"
                                        }
                                    ]
                                }
                            },
                            {
                                "downwardAPI": {
                                    "items": [
                                        {
                                            "path": "namespace",
                                            "fieldRef": {
                                                "apiVersion": "v1",
                                                "fieldPath": "metadata.namespace"
                                            }
                                        }
                                    ]
                                }
                            }
                        ],
                        "defaultMode": 420
                    }
                }
            ],
            "initContainers": [
                {
                    "name": "init",
                    "image": "public.ecr.aws/opsterio/busybox:1.27.2-buildx",
                    "command": [
                        "sh",
                        "-c"
                    ],
                    "args": [
                        "chown -R 1000:1000 /usr/share/opensearch/data"
                    ],
                    "resources": {},
                    "volumeMounts": [
                        {
                            "name": "data",
                            "mountPath": "/usr/share/opensearch/data"
                        },
                        {
                            "name": "kube-api-access-h2thl",
                            "readOnly": true,
                            "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
                        }
                    ],
                    "terminationMessagePath": "/dev/termination-log",
                    "terminationMessagePolicy": "File",
                    "imagePullPolicy": "IfNotPresent",
                    "securityContext": {
                        "runAsUser": 0
                    }
                }
            ],
            "containers": [
                {
                    "name": "opensearch",
                    "image": "docker.io/opensearchproject/opensearch:2.4.1",
                    "command": [
                        "/bin/bash",
                        "-c",
                        "./opensearch-docker-entrypoint.sh"
                    ],
                    "ports": [
                        {
                            "name": "http",
                            "containerPort": 9200,
                            "protocol": "TCP"
                        },
                        {
                            "name": "transport",
                            "containerPort": 9300,
                            "protocol": "TCP"
                        }
                    ],
                    "env": [
                        {
                            "name": "cluster.initial_master_nodes",
                            "value": "my-first-cluster-bootstrap-0"
                        },
                        {
                            "name": "discovery.seed_hosts",
                            "value": "my-first-cluster-discovery"
                        },
                        {
                            "name": "cluster.name",
                            "value": "my-first-cluster"
                        },
                        {
                            "name": "network.bind_host",
                            "value": "0.0.0.0"
                        },
                        {
                            "name": "network.publish_host",
                            "valueFrom": {
                                "fieldRef": {
                                    "apiVersion": "v1",
                                    "fieldPath": "metadata.name"
                                }
                            }
                        },
                        {
                            "name": "OPENSEARCH_JAVA_OPTS",
                            "value": "-Xmx512M -Xms512M -Dopensearch.transport.cname_in_publish_address=true"
                        },
                        {
                            "name": "node.roles",
                            "value": "cluster_manager,data"
                        },
                        {
                            "name": "http.port",
                            "value": "9200"
                        }
                    ],
                    "resources": {
                        "limits": {
                            "cpu": "500m",
                            "memory": "2Gi"
                        },
                        "requests": {
                            "cpu": "500m",
                            "memory": "2Gi"
                        }
                    },
                    "volumeMounts": [
                        {
                            "name": "data",
                            "mountPath": "/usr/share/opensearch/data"
                        },
                        {
                            "name": "admin-credentials",
                            "mountPath": "/mnt/admin-credentials"
                        },
                        {
                            "name": "kube-api-access-h2thl",
                            "readOnly": true,
                            "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
                        }
                    ],
                    "livenessProbe": {
                        "tcpSocket": {
                            "port": 9200
                        },
                        "initialDelaySeconds": 10,
                        "timeoutSeconds": 5,
                        "periodSeconds": 20,
                        "successThreshold": 1,
                        "failureThreshold": 10
                    },
                    "readinessProbe": {
                        "exec": {
                            "command": [
                                "/bin/bash",
                                "-c",
                                "curl -k -u \"$(cat /mnt/admin-credentials/username):$(cat /mnt/admin-credentials/password)\" --silent --fail https://localhost:9200"
                            ]
                        },
                        "initialDelaySeconds": 60,
                        "timeoutSeconds": 30,
                        "periodSeconds": 30,
                        "successThreshold": 1,
                        "failureThreshold": 5
                    },
                    "startupProbe": {
                        "tcpSocket": {
                            "port": 9200
                        },
                        "initialDelaySeconds": 10,
                        "timeoutSeconds": 5,
                        "periodSeconds": 20,
                        "successThreshold": 1,
                        "failureThreshold": 10
                    },
                    "terminationMessagePath": "/dev/termination-log",
                    "terminationMessagePolicy": "File",
                    "imagePullPolicy": "IfNotPresent"
                }
            ],
            "restartPolicy": "Always",
            "terminationGracePeriodSeconds": 30,
            "dnsPolicy": "ClusterFirst",
            "serviceAccountName": "default",
            "serviceAccount": "default",
            "nodeName": "ip-192-168-167-129.ap-southeast-2.compute.internal",
            "securityContext": {},
            "hostname": "my-first-cluster-nodes-0",
            "subdomain": "my-first-cluster",
            "affinity": {
                "podAntiAffinity": {
                    "preferredDuringSchedulingIgnoredDuringExecution": [
                        {
                            "weight": 100,
                            "podAffinityTerm": {
                                "labelSelector": {
                                    "matchExpressions": [
                                        {
                                            "key": "opster.io/opensearch-nodepool",
                                            "operator": "In",
                                            "values": [
                                                "nodes"
                                            ]
                                        }
                                    ]
                                },
                                "topologyKey": "topology.kubernetes.io/zone"
                            }
                        }
                    ]
                }
            },
            "schedulerName": "default-scheduler",
            "tolerations": [
                {
                    "key": "node.kubernetes.io/not-ready",
                    "operator": "Exists",
                    "effect": "NoExecute",
                    "tolerationSeconds": 300
                },
                {
                    "key": "node.kubernetes.io/unreachable",
                    "operator": "Exists",
                    "effect": "NoExecute",
                    "tolerationSeconds": 300
                }
            ],
            "priority": 0,
            "enableServiceLinks": true,
            "preemptionPolicy": "PreemptLowerPriority",
            "topologySpreadConstraints": [
                {
                    "maxSkew": 1,
                    "topologyKey": "topology.kubernetes.io/zone",
                    "whenUnsatisfiable": "DoNotSchedule",
                    "labelSelector": {
                        "matchLabels": {
                            "opster.io/opensearch-cluster": "my-first-cluster",
                            "opster.io/opensearch-nodepool": "nodes"
                        }
                    }
                },
                {
                    "maxSkew": 1,
                    "topologyKey": "kubernetes.io/hostname",
                    "whenUnsatisfiable": "DoNotSchedule",
                    "labelSelector": {
                        "matchLabels": {
                            "opster.io/opensearch-cluster": "my-first-cluster",
                            "opster.io/opensearch-nodepool": "nodes"
                        }
                    }
                }
            ]
        },
        "status": {
            "phase": "Running",
            "conditions": [
                {
                    "type": "Initialized",
                    "status": "True",
                    "lastProbeTime": null,
                    "lastTransitionTime": "2023-04-19T08:53:53Z"
                },
                {
                    "type": "Ready",
                    "status": "True",
                    "lastProbeTime": null,
                    "lastTransitionTime": "2023-04-19T08:55:24Z"
                },
                {
                    "type": "ContainersReady",
                    "status": "True",
                    "lastProbeTime": null,
                    "lastTransitionTime": "2023-04-19T08:55:24Z"
                },
                {
                    "type": "PodScheduled",
                    "status": "True",
                    "lastProbeTime": null,
                    "lastTransitionTime": "2023-04-19T08:53:42Z"
                }
            ],
            "hostIP": "192.168.167.129",
            "podIP": "192.168.163.58",
            "podIPs": [
                {
                    "ip": "192.168.163.58"
                }
            ],
            "startTime": "2023-04-19T08:53:42Z",
            "initContainerStatuses": [
                {
                    "name": "init",
                    "state": {
                        "terminated": {
                            "exitCode": 0,
                            "reason": "Completed",
                            "startedAt": "2023-04-19T08:53:53Z",
                            "finishedAt": "2023-04-19T08:53:53Z",
                            "containerID": "containerd://55cac1d4e3b06f9ee511927bff6683859f99dba539536bed33c85a99219b6e51"
                        }
                    },
                    "lastState": {},
                    "ready": true,
                    "restartCount": 0,
                    "image": "public.ecr.aws/opsterio/busybox:1.27.2-buildx",
                    "imageID": "public.ecr.aws/opsterio/busybox@sha256:8a8ff0bf9d1600ba0b13507a9547f03962f66c362b537e12a950f27d7b57e73e",
                    "containerID": "containerd://55cac1d4e3b06f9ee511927bff6683859f99dba539536bed33c85a99219b6e51"
                }
            ],
            "containerStatuses": [
                {
                    "name": "opensearch",
                    "state": {
                        "running": {
                            "startedAt": "2023-04-19T08:53:53Z"
                        }
                    },
                    "lastState": {},
                    "ready": true,
                    "restartCount": 0,
                    "image": "docker.io/opensearchproject/opensearch:2.4.1",
                    "imageID": "docker.io/opensearchproject/opensearch@sha256:cf80040045595329691a966fe47dedc49dacd262260855dcb768ffce297abed2",
                    "containerID": "containerd://2cc4bcbe7b4871a61dc938645615df3d3150e62bd20ea1f0998f127ff3a2755b",
                    "started": true
                }
            ],
            "qosClass": "Burstable"
        }
    },
    "requestReceivedTimestamp": "2023-04-19T09:10:45.369440Z",
    "stageTimestamp": "2023-04-19T09:10:45.379147Z",
    "annotations": {
        "authorization.k8s.io/decision": "allow",
        "authorization.k8s.io/reason": "RBAC: allowed by ClusterRoleBinding \"opensearch-operator-manager-rolebinding\" of ClusterRole \"opensearch-operator-manager-role\" to ServiceAccount \"opensearch-operator-controller-manager/default\""
    }
}`


This bit seems to be the relevant info:

`        "authorization.k8s.io/reason": "RBAC: allowed by ClusterRoleBinding \"opensearch-operator-manager-rolebinding\" of ClusterRole \"opensearch-operator-manager-role\" to ServiceAccount \"opensearch-operator-controller-manager/default\""`

Which suggests the Operator is in fact telling K8s to terminate the pod for some reason.

**Comment by ccben87** [comment]
I'm happy to do any further troubleshooting to assist in resolving this but am at a bit of a loss of what I can do. Seems clear the Operator is triggering it but no clue why. Is there a way to increase the verbosity of the debug logs?

**Comment by ccben87** [comment]
Okay, so I've gotten closer to figuring out this one. 

When running Operator 2.3.0 on Opensearch 2.4.1 and monitoring the updatedReplicas value for the Node pool's statefulset I get the following:

`  updatedReplicas: 3

  updatedReplicas: 2

  updatedReplicas: 1`

  
  When running Operator 2.3.0 on Opensearch 2.4.0 and monitoring the updatedReplicas value for the Node pool's statefulset I get the following:
  
  `  updatedReplicas: 3

  updatedReplicas: 2

  updatedReplicas: 2

  updatedReplicas: 3`
  
 I'm not sure why but for some reason K8s is incorrectly detecting the replicas as having been updated which causes the Operator via the function https://github.com/Opster/opensearch-k8s-operator/blob/e41413a40c16af732f6d95f3676ae52c129670f8/opensearch-operator/pkg/builders/cluster.go#L875 to issue a 'delete' for all the pods in the set.

**Comment by ccben87** [comment]
Is anyone able to assist in working out this issue? It's holding me back from using this operator as I can't afford my Opensearch cluster to fail any time one of my hosts fails.

**Comment by swoehrl-mw** [comment]
Hi @ccben87 . Thank you for debugging this probem and providing so much information.
I managed to reproduce the issue. As you stated, K8s seems to incorrectly count the terminating pod as not updated, which triggers the rolling restart logic.
I've written a possible fix in PR #545. Could you verify if with this change the problem no longer occurs on your installation?

**Comment by ccben87** [comment]
Hi, thanks for taking a look at this. I am no longer able to reproduce this. This seems to have resolved itself, likely with an update to the EKS node version that automatically happened for me. It would be good to have this PR merged though as an added check to ensure this doesn't happen again.
